sql/core/src/main/scala/org/apache/spark/sql/execution/basicOperators.scala

```diff
@@ -139,9 +139,10 @@ case class TakeOrdered(limit: Int, sortOrder: Seq[SortOrder], child: SparkPlan)
 
   private def collectData(): Array[Row] = child.execute().map(_.copy()).takeOrdered(limit)(ord)
 
-  // TODO: Is this copying for no reason?
-  override def executeCollect(): Array[Row] =
-    collectData().map(ScalaReflection.convertRowToScala(_, this.schema))
+  override def executeCollect(): Array[Row] = {
+    val converter = CatalystTypeConverters.createToScalaConverter(schema)
+    collectData().map(converter(_).asInstanceOf[Row])
```

Comment:
`createToScalaConverter` is used recursively, so it will be generating functions that return Rows only at the top level, while there might be calls that generate functions that return arrays or decimals. Previously I used `convertRowWithConverters`, which was explicitly returning Row, but required an extra line to generate converters. This is cleaner, but I need a cast here.